### PR TITLE
added snake game

### PR DESCRIPTION
**Pull request by b7g**
snake game
# Description
this pr adds a snake game

## Type of change
- [X] New feature (non breaking change that adds functionality)

## Testing
seems functional

# Checklist
- [x] My code follows the general style guidelines of the project
- [x] I have perfomed a self-review of my code
- [x] I have commented my code, particulary where it is unclear
- [x] My changes generate no new warnings or errors
- [x] The project compiles and runs correctly

### CODE REVIEW

File: game/games/snake/canvas.gd [hunk 1]
```diff
@@ -0,0 +1,166 @@
...
+	_set_game_paused(true)
+	randomize()
+	_calc_tile_size()
+	_set_playfield()
+	_set_snake()
+	_set_apple()
+
+
+func _process(_delta):
+	_tick += 1
```
Comment (Numenter) [inline game/games/snake/canvas.gd]: Snake speed varies with FPS. Incorporate delta time into your tick rate.
For testing set **Settings/Debug/Settings/Force FPS** or **Settings/Display/Window/Use Vsync**.

File: game/games/snake/canvas.gd [hunk 2]
```diff
@@ -0,0 +1,166 @@
...
+
+func _game_over():
+	_set_game_paused(true)
+	_set_snake()
+	if _score > _high_score:
+		_high_score = _score
+	_score = 0
+	_main.set_score(_score)
+	_main.set_highscore(_high_score)
+	GameManager.end_game("Ouch! You got a score of %s!" % _high_score, _high_score)
```
Comment (Joshix-1) [inline game/games/snake/canvas.gd]: why do you set the score as _high_score

File: game/games/snake/snake.gd [hunk 3]
```diff
@@ -0,0 +1,20 @@
...
+
+func _ready():
+	var highscore_dict = GameManager.get_highscore()
+	if highscore_dict.has("score"):
+		var highscore = highscore_dict["score"]
+		if highscore != null:
+			set_highscore(highscore)
+
+
+func set_score(score: int):
```
Comment (Joshix-1) [inline game/games/snake/snake.gd]: I would name this display_score

File: game/games/snake/snake.gd [hunk 4]
```diff
@@ -0,0 +1,20 @@
...
+	_label_score.text = "Score: %s" % score
+
+
+func set_highscore(highscore: int):
```
Comment (Joshix-1) [inline game/games/snake/snake.gd]: I would name this display_highscore

File: game/games/snake/canvas.gd [hunk 5]
```diff
@@ -0,0 +1,166 @@
...
+		_score += 1
+		_main.set_score(_score)
+	else:
+		_snake.remove(0)
+
+
+func _draw_snake():
+	var color = Color(0.016, 1, 0)
+	for snake_part in _snake:
+		_draw_square(snake_part, color)
```
Comment (Joshix-1) [inline game/games/snake/canvas.gd]: I think the head should have a different colour

File: game/games/snake/canvas.gd [hunk 6]
```diff
@@ -0,0 +1,166 @@
...
+	if new_head_pos.is_equal_approx(_apple_pos):
+		_set_apple()
+		_score += 1
+		_main.set_score(_score)
+	else:
+		_snake.remove(0)
+
+
+func _draw_snake():
+	var color = Color(0.016, 1, 0)
```
Comment (Joshix-1) [inline game/games/snake/canvas.gd]: why not make this a constant?

File: game/games/snake/canvas.gd (hunk 1 above)
Comment (Joshix-1) [reply game/games/snake/canvas.gd]: yes, something like the following:
```py
_tick += _delta
if _tick >= 0.1:
    _tick -= 0.1
    ...
```

File: game/games/snake/canvas.gd [hunk 7]
```diff
@@ -0,0 +1,166 @@
...
+
+
+func _set_snake():
+	_snake.clear()
+	var center = int(tile_count / 2)
+	_snake.push_back(Vector2(center, center))
+	_direction = null
+
+
+func _set_apple():
```
Comment (Joshix-1) [inline game/games/snake/canvas.gd:100]: What if the apple spawns inside the snake?

Comment (b7g) [reply game/games/snake/canvas.gd:100]: then it is drawn on top on the snake

File: game/games/snake/canvas.gd (hunk 5 above)
Comment (b7g) [reply game/games/snake/canvas.gd]: I don’t see why

Comment (Joshix-1) [reply game/games/snake/canvas.gd]: `_draw_square(_snake.back(), Color(0, 0.5, 0))`
makes it look better

Comment (b7g) [reply game/games/snake/canvas.gd]: It looks like a bending matchstick

File: game/games/snake/canvas.gd [hunk 8]
```diff
@@ -0,0 +1,167 @@
...
+var _direction = Direction.RIGHT
+var _apple_pos := Vector2()
+
+var _score: int = 0
+var _high_score: int = 0
+
+var _collected_delta: float = 0.0
+var _game_paused: bool = false
+
+onready var _main = get_parent().get_parent().get_parent().get_parent()
```
Comment (Numenter) [inline game/games/snake/canvas.gd]:
```suggestion
onready var _main = get_tree().current_scene
```

Comment (b7g) [reply game/games/snake/canvas.gd]: good suggestion, thanks

File: game/games/snake/canvas.gd (hunk 5 above)
Comment (Joshix-1) [reply game/games/snake/canvas.gd]: then what about a gradient on the snake?
e.g. something like the following:
```gd
func _draw_snake():
	var i := 0.0
	for snake_part in _snake:
		i += 0.5 / _snake.size()
		_draw_square(snake_part, Color(0.016, 0.5 + i, 0))
```

For me personally the snake currently looks boring

File: game/games/snake/canvas.gd [hunk 9]
```diff
@@ -0,0 +1,167 @@
...
+	randomize()
+	_calc_tile_size()
+	_set_playfield()
+	_set_snake()
+	_set_apple()
+
+
+func _process(_delta):
+	_collected_delta += _delta
+	while _collected_delta >= 0.067:
```
Comment (Joshix-1) [inline game/games/snake/canvas.gd:37]: For me personally 0.067 is too fast, I like 0.07 better

**Pull request merged**